File: others/Priority Scheduling.c
// Priority Scheduling (Non-Preemptive)
#include<stdio.h>
void swap(int * a, int * b)
{
	int tmp = *a;
	*a = *b;
	*b = tmp;
}
int main()
{
    int n;
    printf("Enter number of processes: ");
    scanf("%d",&n);
    int pid[n], pr[n], bt[n], wt[n], tat[n];
    printf("Enter values: -\n");
    printf("Process ID\tPriority\tBurst Time\t\n");
    for(int i = 0; i < n; i++)
    {
		scanf("%d%d%d",&pid[i],&pr[i],&bt[i]);
    }
    for(int i = 0; i < n; i++)
	{
		for(int j = i; j < n; j++)
		{
			if(pr[i]<pr[j])
			{
                swap(&pr[i], &pr[j]);
				swap(&bt[i], &bt[j]);
				swap(&pid[i], &pid[j]);
			}
		}
	}
	int gc[n+1];
	gc[0] = 0;  // We assume that arrival time is zero
	for(int i = 1; i < n+1; i++)
	{
		gc[i] = gc[i-1] + bt[i-1];
	}
	for(int i = 0; i < n; i++)
	{
		wt[i] = gc[i];
	}
	for(int i = 1; i < n+1; i++)
	{
		tat[i-1] = gc[i];
	}
	int sum_wt = 0, sum_tat = 0;
	float avg_wt, avg_tat;
	for(int i = 0; i < n; i++)
	{
		sum_wt += wt[i];
	}
	avg_wt = (float)sum_wt/n;
	printf("Average waiting time is %f\n",avg_wt);
	for(int i = 0; i < n; i++)
	{
		sum_tat += tat[i];
	}
	avg_tat = (float)sum_tat/n;
	printf("Average turn around time is %f\n",avg_tat);
}
/*

 	Output: -
	Enter number of processes: 3
	Enter values: -
	Process ID	Priority	Burst Time	
	1		2		5
	2		5		4
	3		1		6
	Average waiting time is 4.333333
	Average turn around time is 9.333333

 */

File: others/Priority Scheduling.py
#! /usr/bin/python3

# Priority Scheduling (Non-preemptive Algorithm)

pid = ["P1","P2","P3","P4","P5","P6","P7"]
arrival_time = [0,2,1,4,6,5,7]
burst_time = [3,5,4,2,9,4,10]
priority = [2,6,3,5,7,4,10]


"""

Here, we sort the process according to priority. If the priority is same then sorted according to arrival time scheduling.

"""
# sorting according to arrival time, but if arrival time are same then sorting according to priority
for i in range(len(pid)):
	for j in range(i,len(pid)):
		if arrival_time[i] > arrival_time[j]:
			priority[i], priority[j] = priority[j], priority[i]
			pid[i], pid[j] = pid[j], pid[i]
			arrival_time[i], arrival_time[j] = arrival_time[j], arrival_time[i]
			burst_time[i], burst_time[j] = burst_time[j], burst_time[i]
		elif arrival_time[i] == arrival_time[j] and priority[i] < priority[j]:
			priority[i], priority[j] = priority[j], priority[i]
			pid[i], pid[j] = pid[j], pid[i]
			arrival_time[i], arrival_time[j] = arrival_time[j], arrival_time[i]
			burst_time[i], burst_time[j] = burst_time[j], burst_time[i]

# sorting according to priority, but if priority are same then sorting according to arrival time
for i in range(1, len(pid)):
	for j in range(i, len(pid)):
		if priority[i] > priority[j]:
			priority[i], priority[j] = priority[j], priority[i]
			pid[i], pid[j] = pid[j], pid[i]
			arrival_time[i], arrival_time[j] = arrival_time[j], arrival_time[i]
			burst_time[i], burst_time[j] = burst_time[j], burst_time[i]
		elif priority[i] == priority[j] and arrival_time[i] > arrival_time[j]:
			priority[i], priority[j] = priority[j], priority[i]
			pid[i], pid[j] = pid[j], pid[i]
			arrival_time[i], arrival_time[j] = arrival_time[j], arrival_time[i]
			burst_time[i], burst_time[j] = burst_time[j], burst_time[i]
"""

Now, we apply First Come First Serve

"""

completion_time = []

for i in range(len(pid)):
	if completion_time == []:										# time completion of an initial process
		completion_time.append(arrival_time[i] + burst_time[i])
	elif completion_time[i-1] >= arrival_time[i-1]:					# if any process arrives before the completion of an initial process
		completion_time.append(completion_time[i-1] + burst_time[i])
	else:															# if any process arrives after the completion of an initial process
		completion_time.append(arrival_time[i] + burst_time[i])

turn_around_time = []
waiting_time = []

for i in range(len(pid)):
	turn_around_time.append(completion_time[i] - arrival_time[i])
	waiting_time.append(turn_around_time[i] - burst_time[i])

print("Average Turn Around Time: ",sum(turn_around_time)/len(pid))
print("Average Waiting Time: ",sum(waiting_time)/len(pid))
